jwt/verifyEdDsa.js

```diff
@@ -12,12 +12,12 @@ module.exports = function verifyEdDsa(jsonWebToken, publicKey) {
   if (decodedJwt.header.alg !== 'EdDSA') {
     throw new JlincJwtError('header does not indicate EdDSA');
   }
-  if (publicKey && isB64.test(publicKey)) {
+  if (publicKey !== undefined) { // if publicKey is supplied it must be used!
     pub.key = publicKey;
-  } else if (decodedJwt.header.jwk && decodedJwt.header.jwk.x && isB64.test(decodedJwt.header.jwk.x)) {
+  } else if (decodedJwt.header.jwk && decodedJwt.header.jwk.x && isB64.test(decodedJwt.header.jwk.x)) { // otherwise see if there is a JSON WebKey in the header
     pub.key = decodedJwt.header.jwk.x;
   } else {
-    throw new JlincJwtError('cannot find valid public key');
+    throw new JlincJwtError('cannot find a public key');
```

Comment:
What do you think about?:

```js
'use strict';

const b64 = require('urlsafe-base64');
const sodium = require('sodium').api;

module.exports = function verifyEdDsa(jsonWebToken, publicKey) {
  const { JlincJwtError } = this;

  const decodedJwt = this.read(jsonWebToken);

  if (!decodedJwt || !decodedJwt.header || !decodedJwt.header.alg || !decodedJwt.header.jwk)
    throw new JlincJwtError(`jsonWebToken decoded into an unexpected object`);

  if (decodedJwt.header.alg !== 'EdDSA')
    throw new JlincJwtError('header does not indicate EdDSA');

  let publicKeyDecoded
  if (typeof publicKey !== 'undefined'){ // if publicKey is supplied it must be used!
    publicKeyDecoded = decodePublicKey(publicKey)
    if (!publicKeyDecoded) throw new JlincJwtError(`publicKey is not a valid public key`);
  }else{ // otherwise see if there is a JSON WebKey in the header
    const jsonWebKey = decodedJwt.header.jwk && decodedJwt.header.jwk.x
    if (jsonWebKey){
      publicKeyDecoded = decodePublicKey(publicKey)
      if (!publicKeyDecoded) throw new JlincJwtError(`invalid JSON WebKey in header`);
    }
  }
  
  if (!publicKeyDecoded) throw new JlincJwtError('cannot find a public key');

  const signatureDecoded = b64.decode(decodedJwt.signature);
  const message = Buffer.from(decodedJwt.signed);
  
  let verified
  try {
    verified = sodium.crypto_sign_verify_detached(
      signatureDecoded, message, publicKeyDecoded
    )
  }catch(error){
    throw new JlincJwtError(`EdDsa verification failed: ${error.message}`);
  }
  if (verified) return decodedJwt;
  else throw new JlincJwtError(`invalid signature`);
};

const decodePublicKey = publicKey => {
  if (typeof publicKey !== 'string') return
  const decoded = b64.decode(publicKey)
  if (!decoded || decoded.toString().length < 32) return
  return decoded
}
```

Reply:
All good except it would be sexier (um, I mean more precise) to say
if (!decoded || decoded.length !== sodium.crypto_sign_PUBLICKEYBYTES) return